src/Middleware/ThrottleMiddleware.php

```diff
@@ -0,0 +1,218 @@
+     * @param callable $next Next class in middelware
+     * @return ResponseInterface
+     */
+    public function __invoke(ServerRequestInterface $request, ResponseInterface $response, callable $next)
+    {
+        $this->_setIdentifier($request);
+        $this->_initCache();
+        $this->_count = $this->_touch();
+
+        if ($this->_count > $this->getConfig('limit')) {
+            $stream = new Stream('php://memory', 'wb+');
+            $stream->write((string)$this->getConfig('message'));
+
+            return $response->withStatus(429)
+                ->withBody($stream);
+        }
+
+        $response = $next($request, $response);
+
+        return $this->_setHeaders($response);
+    }
+
+    /**
+     * Sets the identifier class property. Uses Throttle default IP address
+     * based identifier unless a callable alternative is passed.
+     *
+     * @param \Psr\Http\Message\ResponseInterface $request RequestInterface instance
+     * @return void
+     * @throws \InvalidArgumentException
+     */
+    protected function _setIdentifier(RequestInterface $request)
+    {
+        $key = $this->getConfig('identifier');
+        if (!is_callable($this->getConfig('identifier'))) {
+            throw new \InvalidArgumentException('Throttle identifier option must be a callable');
+        }
+        $this->_identifier = $key($request);
+    }
+
+    /**
+     * Initializes cache configuration.
+     *
+     * @return void
+     */
+    protected function _initCache()
+    {
+        if (!Cache::getConfig(static::$cacheConfig)) {
+            Cache::setConfig(static::$cacheConfig, [
+                'className' => $this->_getDefaultCacheConfigClassName(),
+                'prefix' => static::$cacheConfig . '_' . $this->_identifier,
+                'duration' => $this->getConfig('interval'),
+            ]);
+        }
+    }
+
+    /**
+     * Gets the className of the default CacheEngine so the Throttle cache
+     * config can use the same. String cast is required to catch a DebugEngine
+     * array/object for users with DebugKit enabled.
+     *
+     * @return string ClassName property of default Cache engine
+     */
+    protected function _getDefaultCacheConfigClassName()
+    {
+        $config = Cache::getConfig('default');
+        $engine = (string)$config['className'];
+
+        // short cache engine names can be returned immediately
+        if (strpos($engine, '\\') === false) {
+            return $engine;
+        }
+        // fully namespace cache engine names need extracting class name
+        preg_match('/.+\\\\(.+)Engine/', $engine, $matches);
+
+        return $matches[1];
+    }
+
+    /**
+     * Atomically updates cache using default CakePHP increment offset 1.
+     *
+     * Please note that the cache key needs to be initialized to prevent
+     * increment() failing on 0. A separate cache key is created to store
+     * the interval expiration time in epoch.
+     *
+     * @return \Cake\Cache\Cache
```

Comment:
What return type would this be? It looks as if that is exactly how it is returned??